security/rbac/engine/engine.go

```diff
@@ -0,0 +1,187 @@
+	for policyName, condition = range engine.conditions {
+		match, err := matchesCondition(condition, args)
+		if err != nil {
+			unknownPolicies += policyName + ", "
+		}
+		if match {
+			return
+		}
+	}
+	policyName = ""
+	return
+}
+
+// CelEvaluationEngine is the struct for CEL engine.
+type CelEvaluationEngine struct {
+	engines []rbacEngine
+}
+
+// NewCelEvaluationEngine builds a cel evaluation engine from a list of Envoy RBACs.
+func NewCelEvaluationEngine(rbacs []pb.RBAC) (CelEvaluationEngine, error) {
+	if len(rbacs) < 1 || len(rbacs) > 2 {
+		return CelEvaluationEngine{}, status.Errorf(codes.InvalidArgument, "must provide 1 or 2 RBACs")
+	}
+	if len(rbacs) == 2 && (rbacs[0].Action != pb.RBAC_DENY || rbacs[1].Action != pb.RBAC_ALLOW) {
+		return CelEvaluationEngine{}, status.Errorf(codes.InvalidArgument, "when providing 2 RBACs, must have 1 DENY and 1 ALLOW in that order")
+	}
+	var engines []rbacEngine
+	for _, rbac := range rbacs {
+		engines = append(engines, newRbacEngine(rbac))
+	}
+	return CelEvaluationEngine{engines}, nil
+}
+
+// Evaluate is the core function that evaluates whether an RPC is authorized.
+//
+// ALLOW policy. If one of the RBAC conditions is evaluated as true, then the
+// CEL engine evaluation returns allow. If all of the RBAC conditions are
+// evaluated as false, then it returns deny. Otherwise, some conditions are
+// false and some are unknown, it returns undecided.
+//
+// DENY policy. If one of the RBAC conditions is evaluated as true, then the
+// CEL engine evaluation returns deny. If all of the RBAC conditions are
+// evaluated as false, then it returns allow. Otherwise, some conditions are
+// false and some are unknown, it returns undecided.
+//
+// DENY policy + ALLOW policy. If one of the conditions in the DENY policy is
+// evaluated as true, then the CEL engine evaluation returns deny. Otherwise,
+// if one of the conditions in the ALLOW policy is evaluated as true, it
+// returns allow. If all the conditions are evaluated as false, it returns
+// deny. If some conditions are unknown whereas the other conditions are
+// false, it returns undecided.
+func (celEngine CelEvaluationEngine) Evaluate(args AuthorizationArgs) AuthorizationDecision {
+	allUnknownPolicies := ""
+	for _, engine := range celEngine.engines {
+		policyName, unknownPolicies := engine.matches(args)
+		// If any engine matched, return that engine's action.
+		if policyName != "" {
+			var decision Decision
+			if engine.action == pb.RBAC_ALLOW {
+				decision = DecisionAllow
+			} else {
+				decision = DecisionDeny
+			}
+			return AuthorizationDecision{decision, policyName}
+		}
+		allUnknownPolicies += unknownPolicies
+	}
+	// If any engine evaluated to unknown, return unknown.
+	if allUnknownPolicies != "" {
+		return AuthorizationDecision{DecisionUnknown, strings.TrimRight(allUnknownPolicies, ", ")}
+	}
+	// If all engines explicitly did not match.
+	var decision Decision
+	if len(celEngine.engines) == 2 {
```

Comment:
I know we had a New function that did some length checks, but we couldn't simply assume it always got executed as expected, especially since it's a public API(users might misuse it in a multi-goroutine model and somehow did something wrong). Considering a scenario that user somehow creates a celEngine with three rbacEngine(s). In current code, it could be pretty bad since we will return allow for it. 
That's why originally I think we might want to return an error here. An error means something internal is messed up and we are not in a good state for delivering AuthorizationDecision results. Unless you can guarantee that your program can always conduct an AuthorizationDecision, returning an error could be useful IMO. What do you think?

Reply:
You definitely have a point. However, even though the `CelEvaluationEngine` api is public, it consists of unexported field `engines`, which are `rbacEngine`s (also unexported). Thus, for as far as I'm aware, it's actually impossible to create an instance of `CelEvaluationEngine` other than using the factory method, so I think we're fine.
On the other hand, we can maybe check the number of engines being 1 or 2, and just log the error? I am also okay with modifying the function to return error.

Reply:
Yes, let's return an error. I know logically it may look fine, but it doesn't hurt having this check. Otherwise, if it really happens, it's very hard to debug this. It could also be useful to return some useful error information if something bad really happens.

Reply:
If we are using two RBACs, do we also have to check if the RBAC pairs follow the {DENY, ALLOW} order again here?